README.md
# NAME

Catmandu::MediaHaven - Tools to communicate with the Zeticon MediaHaven server

# SYNOPSIS

    use Catmandu::MediaHaven;

    my $mh = Catmandu::MediaHaven->new(
                    url      => '...' ,
                    username => '...' ,
                    password => '...');

    my $result = $mh->search('nature', start => 0 , num => 100);

    die "search failed" unless defined($result);

    for my $res (@{$result->{mediaDataList}}) {
        my $id = $res->{fragmentId};
        my $date = $res->{data};

        print "$id $date\n";
    }

    my $record = $mh->record('q2136s817');
    my $date   = $record->{date};

    print "q2136s817 $date\n";

    $mh->export($id, sub {
       my $data = shift;
       print $data;
    });

# DESCRIPTION

The [Catmandu::MediaHaven](https://metacpan.org/pod/Catmandu::MediaHaven) module is a low end interface to the MediaHaven
REST api. See also: https://archief.viaa.be/mediahaven-rest-api

# METHODS

## new(url => ... , username => ... , password => ...)

Create a new connection to the MediaHaven server.

## search($query, start => ... , num => ...)

Execute a search query against the MediaHaven server and return the result\_list
as a HASH

## record($id)

Retrieve one record from the MediaHaven server based on an identifier. Returns
a HASH of results.

## edit($id,$field,@values)

Edit the metadata of a record

## export($id, $callback)

Export the binary content of a record from the MediaHaven server. The callback
will retrieve a stream of data when the download is available,

# MODULES

[Catmandu::Importer::MediaHaven](https://metacpan.org/pod/Catmandu::Importer::MediaHaven)

[Catmandu::Store::File::MediaHaven](https://metacpan.org/pod/Catmandu::Store::File::MediaHaven)

[Catmandu::Store::File::MediaHaven::Bag](https://metacpan.org/pod/Catmandu::Store::File::MediaHaven::Bag)

[Catmandu::Store::File::MediaHaven::Index](https://metacpan.org/pod/Catmandu::Store::File::MediaHaven::Index)

# AUTHOR

- Patrick Hochstenbach, `<patrick.hochstenbach at ugent.be>`

# LICENSE AND COPYRIGHT

This program is free software; you can redistribute it and/or modify it under the terms
of either: the GNU General Public License as published by the Free Software Foundation;
or the Artistic License.

See [http://dev.perl.org/licenses/](http://dev.perl.org/licenses/) for more information.


Changes
Revision history for Catmandu-MediaHaven

{{$NEXT}}

0.08  2017-11-28 11:21:33 CET
   - adding support for rate limits when exporting data
   
0.07  2017-11-28 10:10:19 CET
   - fixing search result list

0.06  2017-11-21 13:54:35 CET
   - adding more support to change configuration of command line tool

0.05  2017-09-12 11:43:59 CEST
   - adding support for exporters

0.04  2017-09-09 08:30:27 CEST
   - fixing dependencies

0.03  2017-09-07 15:31:57 CEST
   - adding search and sort support

0.02  2017-09-04 16:53:37 CEST
   - fixing importer

0.01  2017-09-04 16:21:13 CEST
   - initial version


lib/Catmandu/Store/File/MediaHaven/Bag.pm
package Catmandu::Store::File::MediaHaven::Bag;

our $VERSION = '0.07';

use Catmandu::Sane;
use Moo;
use Carp;
use Date::Parse;
use POSIX qw(ceil);
use namespace::clean;

with 'Catmandu::Bag';
with 'Catmandu::FileBag';

sub generator {
    my ($self) = @_;

    my $mh  = $self->store->mh;

    my $res = $mh->record($self->name);

    sub {
        state $done = 0;

        return undef if $done;

        $done = 1;

        return $self->_get($res,$res->{originalFileName});
    };
}

sub _get {
    my ($self,$result,$key) = @_;

    my $mh  = $self->store->mh;

    return undef unless $result;

    return undef unless $result->{originalFileName} eq $key;

    my $md5;

    for my $prop (@{$result->{mdProperties}}) {
        if ($prop->{attribute} eq 'md5_viaa') {
            $md5 = $prop->{value};
        }
    }

    return +{
        _id          => $key,
        size         => -1,
        md5          => $md5 ? $md5 : 'none',
        created      => str2time($result->{archiveDate}),
        modified     => str2time($result->{lastModifiedDate}),
        content_type => 'application/zip',
        _stream      => sub {
            my $out   = $_[0];
            my $bytes = 0;
            $mh->export($self->name, sub {
                my $data = shift;
                # Support the Dancer send_file "write" callback
                if ($out->can('syswrite')) {
                    $bytes += $out->syswrite($data) || die "failed to write : $!";
                }
                else {
                    $bytes += $out->write($data) || die "failed to write : $!";;
                }
            });

            $out->close();

            $bytes;
        }
    };
}

sub exists {
    my ($self, $id) = @_;

    my $mh  = $self->store->mh;

    my $res = $mh->record($self->name);

    $res->{originalFileName} eq $id;
}

sub get {
    my ($self, $id) = @_;

    my $mh  = $self->store->mh;

    my $res = $mh->record($self->name);

    return $self->_get($res,$id);
}

sub add {
    my ($self, $data) = @_;
    croak "Add is not supported in the MediaHaven FileStore";
}

sub delete {
    my ($self, $id) = @_;
    croak "Delete is not supported in the MediaHaven FileStore";
}

sub delete_all {
    my ($self) = @_;
    croak "Delete is not supported in the MediaHaven FileStore";
}

sub commit {
    return 1;
}

1;

__END__

=pod

=head1 NAME

Catmandu::Store::File::MediaHaven::Bag - Index of all "files" in a Catmandu::Store::File::MediaHaven "folder"

=head1 SYNOPSIS

    use Catmandu;

    my $store = Catmandu->store('File::MediaHaven' ,
        url => '...' ,
        username => '...' ,
        password => '...' ,
    );

    my $index = $store->index;

    # List all containers
    $index->each(sub {
        my $container = shift;

        print "%s\n" , $container->{_id};
    });

    # Get a folder
    my $folder = $index->get(1234);

    # Get the files in an folder
    my $files = $index->files(1234);

    $files->each(sub {
        my $file = shift;

        my $name         = $file->_id;
        my $size         = $file->size;
        my $content_type = $file->content_type;
        my $created      = $file->created;
        my $modified     = $file->modified;

        $file->stream(IO::File->new(">/tmp/$name"), file);
    });

    # Retrieve a file
    my $file = $files->get("data.dat");

    # Stream a file to an IO::Handle
    $files->stream(IO::File->new(">data.dat"),$file);

=head1 INHERITED METHODS

This Catmandu::Bag implements:

=over 3

=item L<Catmandu::Bag>

=item L<Catmandu::FileBag>

=back


lib/Catmandu/Store/File/MediaHaven/Index.pm
package Catmandu::Store::File::MediaHaven::Index;

our $VERSION = '0.07';

use Catmandu::Sane;
use Moo;
use Carp;
use POSIX qw(ceil);
use Catmandu::Store::File::MediaHaven::Searcher;
use namespace::clean;

with 'Catmandu::Bag';
with 'Catmandu::Searchable';
with 'Catmandu::FileBag::Index';

sub generator {
    my ($self) = @_;

    my $searcher = Catmandu::Store::File::MediaHaven::Searcher->new(
                        bag   => $self ,
                        limit => $self->store->search_limit ,
                        start => 0 ,
                        sort  => $self->store->search_sort ,
                        query => undef ,
                    );

    return $searcher->generator;
}

sub exists {
    my ($self, $id) = @_;

    croak "Need a key" unless defined $id;

    my $res = $self->store->mh->record($id);

    defined($res);
}

sub add {
    my ($self, $data) = @_;

    croak "Add is not supported in the MediaHaven FileStore";
}

sub get {
    my ($self, $id) = @_;

    my $res = $self->store->mh->record($id);

    if ($res) {
        return +{_id => $id};
    }
    else {
        return undef;
    }
}

sub delete {
    my ($self, $id) = @_;

    croak "Delete is not supported in the MediaHaven FileStore";
}

sub delete_all {
    my ($self) = @_;

    croak "Delete is not supported in the MediaHaven FileStore";
}

sub delete_by_query {
    my ($self, %args) = @_;

    croak "Delete is not supported in the MediaHaven FileStore";
}

sub search {
    my ($self, %args) = @_;

    croak "Search is not supported in the MediaHaven FileStore";
}

sub searcher {
    my ($self, %args) = @_;

    Catmandu::Store::File::MediaHaven::Searcher->new(%args, bag => $self);
}

sub count {
    my ($self)    = @_;

    my $mh    = $self->store->mh;

    my $res   = $mh->search(undef);

    $res->{totalNrOfResults};
}

sub commit {
    return 1;
}

1;

__END__

=pod

=head1 NAME

Catmandu::Store::File::MediaHaven::Index - Index of all "Folders" in a MediaHaven database

=head1 SYNOPSIS

    use Catmandu;

    my $store = Catmandu->store('File::MediaHaven' ,
        url => '...' ,
        username => '...' ,
        password => '...' ,
    );

    my $index = $store->index;

    # List all containers
    $index->each(sub {
        my $container = shift;

        print "%s\n" , $container->{_id};
    });

    # Get a folder
    my $folder = $index->get(1234);

    # Get the files in an folder
    my $files = $index->files(1234);

    $files->each(sub {
        my $file = shift;

        my $name         = $file->_id;
        my $size         = $file->size;
        my $content_type = $file->content_type;
        my $created      = $file->created;
        my $modified     = $file->modified;

        $file->stream(IO::File->new(">/tmp/$name"), file);
    });

    # Retrieve a file
    my $file = $files->get("data.dat");

    # Stream a file to an IO::Handle
    $files->stream(IO::File->new(">data.dat"),$file);

=head1 INHERITED METHODS

This Catmandu::Bag implements:

=over 3

=item L<Catmandu::Bag>

=item L<Catmandu::FileBag::Index>

=back


lib/Catmandu/Store/File/MediaHaven/Searcher.pm
package Catmandu::Store::File::MediaHaven::Searcher;

our $VERSION = '0.07';

use Catmandu::Sane;
use Moo;

has bag   => (is => 'ro', required => 1);
has query => (is => 'ro', required => 1);
has start => (is => 'ro', required => 1);
has limit => (is => 'ro', required => 1);
has sort  => (is => 'ro', required => 0);
has total => (is => 'ro');

sub generator {
    my ($self) = @_;

    my $mh    = $self->bag->store->mh;

    my $query = $self->query;
    my $index = $self->start // 0;
    my $limit = $self->limit;
    my $sort  = $self->sort;

    my $res   = $mh->search($query, start => $index, num => $limit, sort => $sort);

    sub {
        state $results         = $res->{mediaDataList};
        state $num_of_results  = $res->{totalNrOfResults};
        state $total           = $self->total;

        if (defined $total) {
            return unless $total;
        }

        if (defined($total) && defined($limit) && $limit > $total) {
            $limit = $total;
        }

        if (@$results > 0) {
            my $hit =  shift @$results;

            $index++;
            $total-- if defined($total);

            return $self->hit2rec($hit);
        }
        elsif ($index < $num_of_results) {
            my $res = $mh->search($query, start => $index, num => $limit, sort => $sort);

            $results = $res->{mediaDataList};

            $total-- if defined($total);

            $index++;

            my $hit = shift @$results;

            return $self->hit2rec($hit);
        }
        return undef;
    };
}

sub hit2rec {
    my ($self,$hit) = @_;

    if ($self->bag->store->id_fixer) {
        return $self->bag->store->id_fixer->fix($hit);
    }
    else {
        my $id = $hit->{fragmentId};
        return +{_id => $id};
    }
}

sub slice {
    my ($self, $start, $total) = @_;
    $start //= 0;
    $self->new(
        bag   => $self->bag,
        query => $self->query,
        start => $self->start + $start,
        limit => $self->limit,
        sort  => $self->sort,
        total => $total,
    );
}

sub count {
    my ($self)    = @_;

    my $mh    = $self->bag->store->mh;

    my $query = $self->query;

    my $res   = $mh->search($query);

    $res->{totalNrOfResults};
}

1;


lib/Catmandu/Store/File/MediaHaven.pm
package Catmandu::Store::File::MediaHaven;

our $VERSION = '0.07';

use Catmandu::Sane;
use Moo;
use Carp;
use Catmandu;
use Catmandu::Store::File::MediaHaven::Index;
use Catmandu::Store::File::MediaHaven::Bag;
use Catmandu::MediaHaven;
use namespace::clean;

with 'Catmandu::FileStore';

has 'url'           => (is => 'ro' , required => 1);
has 'username'      => (is => 'ro' , required => 1);
has 'password'      => (is => 'ro' , required => 1);
has 'record_query'  => (is => 'ro' , default => sub { "q=%%2B(MediaObjectFragmentId:%s)"; });
has 'search_limit'  => (is => 'ro');
has 'search_sort'   => (is => 'ro');
has id_fixer        => (is => 'ro' , init_arg => 'record_id_fix', coerce => sub {Catmandu->fixer($_[0])},);
has 'mh'            => (is => 'lazy');

sub _build_mh {
    my $self = shift;
    return Catmandu::MediaHaven->new(
        url      => $self->url,
        username     => $self->username,
        password     => $self->password,
        record_query => $self->record_query,
    );
}

1;

__END__

=pod

=head1 NAME

Catmandu::Store::File::MediaHaven - A Catmandu::FileStore to access files in the Zeticon MediaHaven server

=head1 SYNOPSIS

    # From the command line

    $ cat catmandu.yml
    ---
    store:
         mh:
             package: MediaHaven
             options:
                 url: https://archief.viaa.be/mediahaven-rest-api/resources/media
                 username: ...
                 password: ...

    # Export a list of all file containers
    $ catmandu export mh to YAML

    # Export a list of all file containers based on a query
    $ catmandu export mh --query "+(MediaObjectFragmentTitle:data)"
    $ catmandu export mh --query "+(MediaObjectFragmentTitle:data)"  --sort "+=MediaObjectFragmentTitle"

    $ Count all file containers
    $ catmandu count mh

    # Export a list of all files in container '1234'
    $ catmandu export mh --bag 1234 to YAML

    # Download the file 'myfile.txt' from the container '1234'
    $ catmandu stream mh --bag 1234 --id myfile.txt to /tmp/output.txt


    # From Perl
    use Catmandu;

    my $store = Catmandu->store('File::MediaHaven' ,
        url      => '...' ,
        username => '...' ,
        password => '...' ,
    );

    my $index = $store->index;

    # List all folder
    $index->bag->each(sub {
        my $container = shift;

        print "%s\n" , $container->{_id};
    });

    # Get the folder
    my $files = $index->files('1234');

    # Retrieve a file
    my $file = $files->get('foobar.txt');

    # Stream the contents of a file
    $files->stream(IO::File->new('>foobar.txt'), $file);

=head1 METHODS

=head2 new(%connection_parameters)

Create a new Catmandu::Store::File::MediaHaven with the following connection
parameters:

=over

=item url

Required. The URL to the MediaHaven REST endpoint.

=item username

Required. Username used to connect to MediaHaven.

=item password

Required. Password used to connect to MediaHaven.

=item record_query

Optional. MediaHaven query to extract a given identifier from the database.
Default: "q=%%2B(MediaObjectFragmentId:%s)"

=item record_id_fix

Optional. One or more L<Catmandu::Fix> commands or a Fix script used to
extract the C<_id> bag identifier from the MediaHaven record.

=item search_limit

Optional. The maximum number of records to return for each Mediahaven search query.

=item search_sort

Optional. The sorting field used in queries.

=back


=head1 INHERITED METHODS

This Catmandu::FileStore implements:

=over 3

=item L<Catmandu::FileStore>

=back

The index Catmandu::Bag in this Catmandu::Store implements:

=over 3

=item L<Catmandu::Bag>

=item L<Catmandu::Searchable>

=item L<Catmandu::FileBag::Index>

=back

The file Catmandu::Bag in this Catmandu::Store implements:

=over 3

=item L<Catmandu::Bag>

=item L<Catmandu::FileBag>

=back

=head1 SEE ALSO

L<Catmandu::Store::File::MediaHaven::Index>,
L<Catmandu::Store::File::MediaHaven::Bag>,
L<Catmandu::FileStore>

=cut
